src/cluster.c

```diff
@@ -98,7380 +49,1439 @@ static void clusterBuildMessageHdr(clusterMsg *hdr, int type, size_t msglen);
-    {CLUSTER_NODE_HANDSHAKE,    "handshake,"},
-    {CLUSTER_NODE_NOADDR,       "noaddr,"},
-    {CLUSTER_NODE_NOFAILOVER,   "nofailover,"}
-};
-
-/* Concatenate the comma separated list of node flags to the given SDS
- * string 'ci'. */
-sds representClusterNodeFlags(sds ci, uint16_t flags) {
-    size_t orig_len = sdslen(ci);
-    int i, size = sizeof(redisNodeFlagsTable)/sizeof(struct redisNodeFlags);
-    for (i = 0; i < size; i++) {
-        struct redisNodeFlags *nodeflag = redisNodeFlagsTable + i;
-        if (flags & nodeflag->flag) ci = sdscat(ci, nodeflag->name);
-    }
-    /* If no flag was added, add the "noflags" special flag. */
-    if (sdslen(ci) == orig_len) ci = sdscat(ci,"noflags,");
-    sdsIncrLen(ci,-1); /* Remove trailing comma. */
-    return ci;
-}
-
-/* Concatenate the slot ownership information to the given SDS string 'ci'.
- * If the slot ownership is in a contiguous block, it's represented as start-end pair,
- * else each slot is added separately. */
-sds representSlotInfo(sds ci, uint16_t *slot_info_pairs, int slot_info_pairs_count) {
-    for (int i = 0; i< slot_info_pairs_count; i+=2) {
-        unsigned long start = slot_info_pairs[i];
-        unsigned long end = slot_info_pairs[i+1];
-        if (start == end) {
-            ci = sdscatfmt(ci, " %i", start);
-        } else {
-            ci = sdscatfmt(ci, " %i-%i", start, end);
-        }
-    }
-    return ci;
-}
-
-/* Generate a csv-alike representation of the specified cluster node.
- * See clusterGenNodesDescription() top comment for more information.
- *
- * The function returns the string representation as an SDS string. */
-sds clusterGenNodeDescription(client *c, clusterNode *node, int use_pport) {
-    int j, start;
-    sds ci;
-    int port = use_pport && node->pport ? node->pport : node->port;
-
-    /* Node coordinates */
-    ci = sdscatlen(sdsempty(),node->name,CLUSTER_NAMELEN);
-    /* Node's ip/port and optional announced hostname */
-    if (sdslen(node->hostname) != 0) {
-        ci = sdscatprintf(ci," %s:%i@%i,%s",
-            node->ip,
-            port,
-            node->cport,
-            node->hostname);
-    } else {
-        ci = sdscatprintf(ci," %s:%i@%i,",
-            node->ip,
-            port,
-            node->cport);
-    }
-
-    /* Don't expose aux fields to any clients yet but do allow them
-     * to be persisted to nodes.conf */
-    if (c == NULL) {
-        for (int i = af_start; i < af_count; i++) {
-            if (auxFieldHandlers[i].isPresent(node)) {
-                ci = sdscatprintf(ci, ",%s=", auxFieldHandlers[i].field);
-                ci = auxFieldHandlers[i].getter(node, ci);
-            }
-        }
-    }
-
-    /* Flags */
-    ci = sdscatlen(ci," ",1);
-    ci = representClusterNodeFlags(ci, node->flags);
-
-    /* Slave of... or just "-" */
-    ci = sdscatlen(ci," ",1);
-    if (node->slaveof)
-        ci = sdscatlen(ci,node->slaveof->name,CLUSTER_NAMELE
```

Comment:
I still want to understand why we are inlining this as opposed to building this as a first party module, similar to what we did with TLS.

Reply:
The more time passes, the less comfortable I get with hard deprecating the existing clustering system. At the very least, making this a module allows someone to still easily integrate it while also getting the latest features.

Reply:
I agree the current cluster is going to be around for a long while, and we should aim for fully modularizing the clustering interface. What are the downsides of doing so?